custom/ilsgateway/api.py

```diff
@@ -316,14 +320,83 @@ def product_sync(self, ilsgateway_product):
+    def _create_web_user(self, email, ilsgateway_webuser, location_id, user_dict):
+        user = WebUser.create(domain=None, username=email,
+                              password=ilsgateway_webuser.password, email=ilsgateway_webuser.email.lower(),
+                              **user_dict)
+        user.add_domain_membership(self.domain, location_id=location_id)
+        dm = user.get_domain_membership(self.domain)
         dm.role_id = UserRole.get_read_only_role_by_domain(self.domain).get_id
-        web_user.save()
-        return web_user
+        user.save()
+        return user
+
+    def _edit_web_user(self, user, location_id):
+        read_only_role = UserRole.get_read_only_role_by_domain(self.domain).get_id
+        if self.domain not in user.get_domains():
+            user.add_domain_membership(self.domain, location_id=location_id, role_id=read_only_role)
+            user.save()
+        else:
+            dm = user.get_domain_membership(self.domain)
+            if dm.role_id != read_only_role or dm.location_id != location_id:
+                dm.role_id = read_only_role
+                dm.location_id = location_id
+                user.save()
+
+    def web_user_sync(self, ilsgateway_webuser):
+        email = ilsgateway_webuser.email.lower()
+        if not email:
+            try:
+                validate_email(ilsgateway_webuser.username)
+                email = ilsgateway_webuser.username
+            except ValidationError:
+                return None
+        user = WebUser.get_by_username(email)
+        user_dict = {
+            'first_name': ilsgateway_webuser.first_name,
+            'last_name': ilsgateway_webuser.last_name,
+            'is_active': ilsgateway_webuser.is_active,
+            'password_hashed': True,
+        }
+        location_id = None
+        if ilsgateway_webuser.location:
+            try:
+                sql_location = SQLLocation.objects.get(domain=self.domain, external_id=ilsgateway_webuser.location)
+                location_id = sql_location.location_id
+            except SQLLocation.DoesNotExist:
+                location_id = None
+
+        if user is None:
+            try:
+                ils_sql_web_user = ILSGatewayWebUser.objects.get(external_id=ilsgateway_webuser.id)
+            except ILSGatewayWebUser.DoesNotExist:
+                ils_sql_web_user = None
+
+            if ils_sql_web_user:
+                # if user exists in db it means he was already migrated but he changed email in v1
+                old_email = ils_sql_web_user.email
+                user = WebUser.get_by_username(old_email)
+                ils_domains = ILSGatewayConfig.get_all_enabled_domains()
+                # make sure it's user migrated from ILS and username is available
+                if all([domain in ils_domains for domain in user.domains])\
+                        and not WebUser.get_by_username(email):
+                    user.username = email
```

Comment:
I'm not sure if I like the idea of updating the user's username since it also will require updating the user's django user entry and then it starts to get messy.  Would it be possible to instead just remove the domain membership from the old user and create a new user? (I believe that's what the ticket specified to do as well)